src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -1,736 +1,755 @@
+
+                    await RefreshIndividualKeyValues(cancellationToken).ConfigureAwait(false);
+                    await RefreshKeyValueCollections(cancellationToken).ConfigureAwait(false);
+                    await RefreshKeyValueAdapters(cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    Interlocked.Exchange(ref _networkOperationsInProgress, 0);
+                }
+            }
+        }
+
+        public async Task<bool> TryRefreshAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await RefreshAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                if (IsAuthenticationError(e))
+                {
+                    _logger?.LogWarning(e, LoggingConstants.RefreshFailedDueToAuthenticationError);
+                }
+                else
+                {
+                    _logger?.LogWarning(e, LoggingConstants.RefreshFailedError);
+                }
+
+                return false;
+            }
+            catch (AggregateException e) when (e?.InnerExceptions?.All(e => e is RequestFailedException) ?? false)
+            {
+                if (IsAuthenticationError(e))
+                {
+                    _logger?.LogWarning(e, LoggingConstants.RefreshFailedDueToAuthenticationError);
+                }
+                else
+                {
+                    _logger?.LogWarning(e, LoggingConstants.RefreshFailedError);
+                }
+
+                return false;
+            }
+            catch (KeyVaultReferenceException e)
+            {
+                _logger?.LogWarning(e, LoggingConstants.RefreshFailedDueToKeyVaultError);
+                return false;
+            }
+            catch (OperationCanceledException)
+            {
+                _logger?.LogWarning(LoggingConstants.RefreshCanceledError);
+                return false;
+            }
+
+            return true;
+        }
+
+        public void SetDirty(TimeSpan? maxDelay)
+        {
+            DateTimeOffset cacheExpires = AddRandomDelay(DateTimeOffset.UtcNow, maxDelay ?? DefaultMaxSetDirtyDelay);
+
+            foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers)
+            {
+                changeWatcher.CacheExpires = cacheExpires;
+            }
+
+            foreach (KeyValueWatcher changeWatcher in _options.MultiKeyWatchers)
+            {
+                changeWatcher.CacheExpires = cacheExpires;
+            }
+        }
+
+        public void ProcessPushNotification(PushNotification notification, TimeSpan? maxDelay)
+        {
+            
+            if (notification == null) 
+            { 
+                throw new ArgumentNullException(nameof(notification)); 
+            }
+
+            if (notification.SyncToken == null || notification.EventType == null || notification.ResourceUri == null)
```

Comment:
The standard way of doing this is

```
if (string.IsNullOrEmpty(propA))
{
    throw new ArgumentNullException(nameof(propA));
}

if (string.IsNullOrEmpty(propB))
{
    throw new ArgumentNullException(nameof(propB));
}

if (string.IsNullOrEmpty(propC))
{
    throw new ArgumentNullException(nameof(propC));
}

//
// Continue for however many properties are required.
```

Reply:
It makes sense if we check the passed arguments this way, but does this standard also apply to the properties of passed arguments?
Say if `notification.SyncToken` is null, does that mean its an `ArgumentNullException` or an `ArgumentException`?

Reply:
ArgumentNullException. ArgumentNullException is just an ArgumentException that's error condition is that the argument is null. So since ArgumentException is being considered, and the issue is that the argument is null, then ArgumentNullException would be best.

Reply:
nameof would use the child property name at the least. `throw new ArgumentNullException(nameof(propC.ChildProperty));` Or possibly `throw new ArgumentNullException($"{nameof(propC)}.{nameof(propC.ChildProperty)}");`

Reply:
Throwing `ArgumentNullException` looks odd to me if the actual argument (`notification` in this case) is not null.
Found [this article](https://social.technet.microsoft.com/wiki/contents/articles/51035.c-proper-use-of-argumentexception-and-argumentnullexception.aspx#:~:text=Basically%2C%20ArgumentException%20is%20deriving%20from%20SystemException%20whereas%20ArgumentNullException,ArgumentException%20is%20more%20generic%2C%20ArgumentNullException%20is%20more%20specific.) as a reference.

Reply:
It does look like the accepted practice here is to use `ArgumentException` for null properties. I ended up on the wrong side of that at some point I suppose. Let's keep ArgumentException then, but let's still do individual validation rather than grouped.